### PR TITLE
Crash support: Restore crash context on filter's posted callback

### LINKED ISSUE #16241: Crash support: Restore crash context on filter's posted callback
Crash support: Restore crash context on filter's posted callback

*Description*:

There exists a variety of filters that stop iteration to make an asynchronous call that on completion calls back the filter. Lets call these *cb_filters*. The issue is that the context used for crash dumping gets lost. Hence if we crashed in this callstack, we’d be missing the crash dumps that we’d usually get.

Proposed solution:
- *cb_filters* can easily reconstruct the crash dump context (the necessary `ScopeTrackedObject`) via a method on the `ActiveStream` / `Filter Manager`.
- If *cb_filters* do not add the context themselves, we complain, and add restore the crash context at a common entry point such as `commonContinue()`.

### PR DESCRIPTION
**Pull request by KBaichoo**
Implemented Crash Support when processing request or response due to a filter's callback.

Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message:Implemented Crash Support when processing request or response due to a filter's callback.
Additional Description:
Risk Level: medium
Testing: Unit and E2E tests
Docs Changes:  NA
Release Notes: Included
Fixes #16241 

Some initial discussion on this PR is here: https://github.com/KBaichoo/envoy/pull/36

### CODE REVIEW

**Comment by antoniovicente** [comment]
kill_request:crash_integration_test seems to be failing due to use-after-free in the gcc build, please take a look since it may be related to this change.

```
2021-05-07T16:28:46.5830625Z [  DEATH   ] pure virtual method called
2021-05-07T16:28:46.5831134Z [  DEATH   ] terminate called without an active exception
```

File: source/common/http/filter_manager.cc [hunk 1]
```diff
@@ -53,6 +56,17 @@ void ActiveStreamFilterBase::commonContinue() {
     return;
   }
 
+  // Set ScopeTrackerScopeState if there's no existing crash context.
+  absl::optional<OpaqueScopeTrackedObject> encapsulated_object;
```
Comment (dmitri-d) [inline source/common/http/filter_manager.cc:60]: Why did you decide to use absl::optional here?

File: source/common/common/opaque_scope_tracker.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+#include "absl/container/inlined_vector.h"
+
+namespace Envoy {
+
+// A small class for encapsulating 1 or more scope tracked objects.
+//
+// This is currently used to restore the underlying request context if the
+// filter continues processing a request due to a callback that it had posted.
+class OpaqueScopeTrackedObject : public ScopeTrackedObject {
```
[the remark this answers is not in the document]
Comment (dmitri-d) [reply source/common/common/opaque_scope_tracker.h:18]: > This object seems somewhat close to the ScopeTrackedObjectStack object

I think they are confusingly close; perhaps `ScopeTrackerScopeState` could be expanded to support multiple objects per tracker (with the addition of changes suggested by @antoniovicente)?

**Review by KBaichoo — COMMENTED** [review]
Thanks for the feedback!

**Comment by KBaichoo** [comment]
/retest 

Didn't restart the coverage test in the prior kick.

**Review by antoniovicente — DISMISSED** [review]
Looks good.  Thanks.

**Comment by KBaichoo** [comment]
Think this is another flake...

File: docs/root/version_history/current.rst [hunk 3]
```diff
@@ -52,6 +52,7 @@ Removed Config or Runtime
 New Features
 ------------
 
+* crash support: restore crash context when processing requests or responses due to a filter's callback if the callback doesn't already restore the context.
```
Comment (htuch) [inline docs/root/version_history/current.rst:55]: Code looks good, but I would really like both here and in interface and other comments below to be more crisp on what a "filter's callback" is. I think this is one of these things that if you've spent a lot of time looking at the problem it's obvious, but to me, there are many callbacks that exist for filters; filters implement callback functions, they invoke callbacks, etc. I think giving an "e.g. foo" would help.

FWIW I think I have some intuition here based on our conversation the other day, but a random reader would not.

Comment (KBaichoo) [reply docs/root/version_history/current.rst]: Done.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by htuch** [comment]
@KBaichoo CI finally passed but needs merge for version history.

**Comment by KBaichoo** [comment]
@htuch merged and CI is now green again.

**Comment by antoniovicente** [comment]
Individual test cases in crash_integration_test take 34 seconds to run.  That doesn't seem right.  Can we look into where the time is going and if we have overly long timeouts?  Sample log below

/wait

```
2021-05-17T16:32:17.0675140Z [2021-05-17 16:24:43.388][30959][trace][connection] [source/common/network/connection_impl.cc:410] [C16] raising connection event 2
2021-05-17T16:32:17.0676111Z [2021-05-17 16:24:43.388][30959][debug][client] [source/common/http/codec_client.cc:88] [C16] connected
2021-05-17T16:32:17.0677214Z [2021-05-17 16:24:43.388][19743][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x2fb5a00 for 3600000ms, min is 3600000ms
2021-05-17T16:32:17.0678338Z [2021-05-17 16:25:17.074][30959][debug][connection] [source/common/network/connection_impl.cc:133] [C16] closing data_to_write=0 type=1
2021-05-17T16:32:17.0679549Z [2021-05-17 16:25:17.074][30959][debug][connection] [source/common/network/connection_impl.cc:243] [C16] closing socket: 1
```

**Comment by KBaichoo** [comment]
The longer runtime above is when running with coverage. 

See: https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/75892/logs/262

These tests execute much faster under normal runs ~13s : https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/75936/logs/68

The tests don't do anything fancy compared to other integration tests apart from death.

WDYT we should do?

**Comment by antoniovicente** [comment]
13 seconds for normal runs vs 420 seconds under coverage is surprising.  It's possible that this is related to death tests and coverage.

I guess we should merge this PR, let me take one final look.

**Review by antoniovicente — APPROVED** [review]

**Pull request merged**